### DISCUSSION TITLE
Secure Zip documentation

### DISCUSSION THREAD

**Discussion by pmqs** [nmoinvaz/minizip#665] [Q&A]
In https://github.com/zlib-ng/minizip-ng/issues/141#issuecomment-484963347 it says this about Secure Zip 

> This is possible now. See How to Create a Secure Zip under Wiki.

Don't see anything in https://github.com/zlib-ng/minizip-ng/wiki

**Comment by nmoinvaz** [comment]
That information is now available in the docs folder.

**Comment by pmqs** [comment]
Thanks @nmoinvaz  

given your use of new headers, I assume the intention with this change is not to make a solution that is binary compatible with the pkzip CD encryption, but to create an equivalent parallel solution to the same issue?

I the author of  [zipdetails](https://github.com/pmqs/zipdetails), so I will want to add support for your new headers at some point. Is the current implementation a work-in-progress, or is it good to go?

A few random observations after a very quick play via the `minizip` commandline

- I can create a non-encrypted CD with the minizip command line (by including the `-z` option) that still has the `encrypted central directory` by setting in the GP flags. Is that a valid use case?
- with the pkzip version, you can tell you have an encrypted CD by the presence of a V2 `Zip64 End of Central Directory Record`. With your solution it looks like the presence of your new headers in a `Central Directory Header` is the equivalent flag.

Once I get some time, I'll dig further into the edge conditions.

**Comment by nmoinvaz** [comment]
You are correct; I did not make a solution that is compatible with PKZIP's CD encryption because that feature has different licensing restrictions that are mentioned in the PKZIP app note.

> I can create a non-encrypted CD with the minizip command line (by including the -z option) that still has the encrypted central directory by setting in the GP flags. Is that a valid use case?

I'm not exactly sure what you mean by "encrypted central directory by setting in the GP flags. Is that a valid use case".

If I remember correctly (it has been a while since I last looked at this stuff), it is a two step process:

* Zipping up the central directory into its own zip entry with the name `__CDCD__`, a new CD is then created that shows only `__CDCD__`.
* Since the CD is now a zip entry, it can be encrypted with a password which must be supplied.

Certificates come into play to validate the hash of the zip entries if that is desired, and that certificate info is stored in extra field data for the zip entry along with the hash info.

**Reply by pmqs** [reply]
> You are correct; I did not make a solution that is compatible with PKZIP's CD encryption because that feature has different licensing restrictions that are mentioned in the PKZIP app note.

that's what I assumed. As far as I can tell, pkzip has the only implementation of their version of CD encryption.

> > I can create a non-encrypted CD with the minizip command line (by including the -z option) that still has the encrypted central directory by setting in the GP flags. Is that a valid use case?
> 
> I'm not exactly sure what you mean by "encrypted central directory by setting in the GP flags. Is that a valid use case".

Consider this -- I can create a non-encrypted CD with the command line `minizip` by including the `-z` option, but not including the `-s` option

```
minizip -0 -z test.zip lorem.txt
```
Below is a dump of the output it created with the `test.zip` file above. The point to note is the `Encrypted Central Dir` bit is set in the first local header but the central directory is not actually encrypted.  That flag being set is not strictly correct.

```
0000 0004 50 4B 03 04 LOCAL HEADER #1       04034B50
0004 0001 14          Extract Zip Spec      14 '2.0'
0005 0001 00          Extract OS            00 'MS-DOS'
0006 0002 08 28       General Purpose Flag  2808
                      [Bit  3]              1 'Streamed'
                      [Bit 11]              1 'Language Encoding'
                      [Bit 13]              1 'Encrypted Central Dir'    <============ NOTE this bit is set
0008 0002 00 00       Compression Method    0000 'Stored'
000A 0004 00 00 00 00 Last Mod Time         00000000 'No Date/Time'
000E 0004 00 00 00 00 CRC                   00000000
0012 0004 00 00 00 00 Compressed Length     00000000
0016 0004 00 00 00 00 Uncompressed Length   00000000
001A 0002 03 00       Filename Length       0003
001C 0002 00 00       Extra Length          0000
001E 0003 30 5F 30    Filename              '0_0'
0021 01BE ...         PAYLOAD

01DF 0004 50 4B 07 08 STREAMING DATA HEADER 08074B50
01E3 0004 00 00 00 00 CRC                   00000000
01E7 0004 BE 01 00 00 Compressed Length     000001BE
01EB 0004 00 00 00 00 Uncompressed Length   00000000

01EF 0004 50 4B 03 04 LOCAL HEADER #2       04034B50
01F3 0001 14          Extract Zip Spec      14 '2.0'
01F4 0001 00          Extract OS            00 'MS-DOS'
01F5 0002 08 08       General Purpose Flag  0808
                      [Bit  3]              1 'Streamed'
                      [Bit 11]              1 'Language Encoding'
01F7 0002 00 00       Compression Method    0000 'Stored'
01F9 0004 2C BE 9D 55 Last Mod Time         559DBE2C 'Thu Dec 29 23:49:24 2022'
01FD 0004 00 00 00 00 CRC                   00000000
0201 0004 00 00 00 00 Compressed Length     00000000
0205 0004 5F 00 00 00 Uncompressed Length   0000005F
0209 0002 08 00       Filename Length       0008
020B 0002 0C 00       Extra Length          000C
020D 0008 5F 5F 63 64 Filename              '__cdcd__'
          63 64 5F 5F
0215 0002 CD CD       Extra ID #0001        CDCD 'Minizip Central Directory'
0217 0002 08 00         Length              0008
0219 0008 01 00 00 00   Entries             0000000000000001
          00 00 00 00

0221 0004 50 4B 01 02 CENTRAL HEADER #1     02014B50
0225 0001 3F          Created Zip Spec      3F '6.3'
0226 0001 03          Created OS            03 'Unix'
0227 0001 14          Extract Zip Spec      14 '2.0'
0228 0001 00          Extract OS            00 'MS-DOS'
0229 0002 08 28       General Purpose Flag  2808
                      [Bit  3]              1 'Streamed'
                      [Bit 11]              1 'Language Encoding'
                      [Bit 13]              1 'Encrypted Central Dir'
022B 0002 00 00       Compression Method    0000 'Stored'
022D 0004 15 8A 99 55 Last Mod Time         55998A15 'Sun Dec 25 17:16:42 2022'
0231 0004 FF E7 0E F9 CRC                   F90EE7FF
0235 0004 BE 01 00 00 Compressed Length     000001BE
0239 0004 BE 01 00 00 Uncompressed Length   000001BE
023D 0002 09 00       Filename Length       0009
023F 0002 28 00       Extra Length          0028
0241 0002 00 00       Comment Length        0000
0243 0002 00 00       Disk Start            0000
0245 0002 00 00       Int File Attributes   0000
                      [Bit 0]               0 'Binary Data'
0247 0004 80 00 F8 81 Ext File Attributes   81F80080
024B 0004 00 00 00 00 Local Header Offset   00000000
024F 0009 6C 6F 72 65 Filename              'lorem.txt'
          6D 2E 74 78
          74
0258 0002 51 1A       Extra ID #0001        1A51 'Minizip Hash'
025A 0002 24 00         Length              0024
025C 0002 17 00         Algorithm           'SHA256'
025E 0002 20 00         Digest Size         0020
0260 0020 90 F2 D6 A2   Digest              90 F2 D6 A2 48 21 C0 71 8E 63 2F 7B
          48 21 C0 71                       FA 70 90 75 E7 E9 CC E0 09 DB 4A 14
          8E 63 2F 7B                       1C 04 07 1B 13 CA FE AC
          FA 70 90 75
          E7 E9 CC E0
          09 DB 4A 14
          1C 04 07 1B
          13 CA FE AC

0280 0004 50 4B 07 08 STREAMING DATA HEADER 08074B50
0284 0004 D7 B9 39 AA CRC                   AA39B9D7
0288 0004 5F 00 00 00 Compressed Length     0000005F
028C 0004 5F 00 00 00 Uncompressed Length   0000005F

0290 0004 50 4B 01 02 CENTRAL HEADER #2     02014B50
0294 0001 3F          Created Zip Spec      3F '6.3'
0295 0001 03          Created OS            03 'Unix'
0296 0001 14          Extract Zip Spec      14 '2.0'
0297 0001 00          Extract OS            00 'MS-DOS'
0298 0002 08 08       General Purpose Flag  0808
                      [Bit  3]              1 'Streamed'
                      [Bit 11]              1 'Language Encoding'
029A 0002 00 00       Compression Method    0000 'Stored'
029C 0004 2C BE 9D 55 Last Mod Time         559DBE2C 'Thu Dec 29 23:49:24 2022'
02A0 0004 D7 B9 39 AA CRC                   AA39B9D7
02A4 0004 5F 00 00 00 Compressed Length     0000005F
02A8 0004 5F 00 00 00 Uncompressed Length   0000005F
02AC 0002 08 00       Filename Length       0008
02AE 0002 34 00       Extra Length          0034
02B0 0002 00 00       Comment Length        0000
02B2 0002 00 00       Disk Start            0000
02B4 0002 00 00       Int File Attributes   0000
                      [Bit 0]               0 'Binary Data'
02B6 0004 00 00 00 00 Ext File Attributes   00000000
02BA 0004 EF 01 00 00 Local Header Offset   000001EF
02BE 0008 5F 5F 63 64 Filename              '__cdcd__'
          63 64 5F 5F
02C6 0002 51 1A       Extra ID #0001        1A51 'Minizip Hash'
02C8 0002 24 00         Length              0024
02CA 0002 17 00         Algorithm           'SHA256'
02CC 0002 20 00         Digest Size         0020
02CE 0020 73 CF 6B B9   Digest              73 CF 6B B9 A9 C0 21 AC F7 56 50 92
          A9 C0 21 AC                       5A 7F 72 C0 BF 76 5C 68 9D D0 7E 60
          F7 56 50 92                       1B B6 53 39 B0 4F F3 13
          5A 7F 72 C0
          BF 76 5C 68
          9D D0 7E 60
          1B B6 53 39
          B0 4F F3 13
02EE 0002 CD CD       Extra ID #0002        CDCD 'Minizip Central Directory'
02F0 0002 08 00         Length              0008
02F2 0008 01 00 00 00   Entries             0000000000000001
          00 00 00 00

02FA 0004 50 4B 05 06 END CENTRAL HEADER    06054B50
02FE 0002 00 00       Number of this disk   0000
0300 0002 00 00       Central Dir Disk no   0000
0302 0002 02 00       Entries in this disk  0002
0304 0002 02 00       Total Entries         0002
0306 0004 6A 00 00 00 Size of Central Dir   0000006A
030A 0004 90 02 00 00 Offset to Central Dir 00000290
030E 0002 00 00       Comment Length        0000
```

**Reply by nmoinvaz** [reply]
The way I read the PKZIP app note is that Bit 13 is used while doing encrypted CD, but it indicates masking of local info. So it doesn't necessary indicate CD encryption, only that select data values are masked to hide their actual values. 
```
Bit 13: Set when encrypting the Central Directory to indicate 
        selected data values in the Local Header are masked to
        hide their actual values.
```
In minizip-ng I use `MZ_ZIP_FLAG_MASK_LOCAL_INFO` for this flag.

**Reply by nmoinvaz** [reply]
I think for PKWARE zip encrypted CD detection, you should probably check for V2 of Zip64 EofCD record, which is greater than V1 in size.

**Reply by pmqs** [reply]
> The way I read the PKZIP app note is that Bit 13 is used while doing encrypted CD, but it indicates masking of local info. So it doesn't necessary indicate CD encryption, only that select data values are masked to hide their actual values.
> 
> ```
> Bit 13: Set when encrypting the Central Directory to indicate 
>         selected data values in the Local Header are masked to
>         hide their actual values.
> ```
> 
> In minizip-ng I use `MZ_ZIP_FLAG_MASK_LOCAL_INFO` for this flag.

I agree

**Reply by pmqs** [reply]
> I think for PKWARE zip encrypted CD detection, you should probably check for V2 of Zip64 EofCD record, which is greater than V1 in size.

Yep, that's what I already use.

**Comment by dgm3333** [comment]
@nmoinvaz above says: That information [How to Create a Secure Zip] is now available in the docs folder.
However the only link I can find just appears to list the actions required.
https://github.com/zlib-ng/minizip-ng/tree/master/doc#how-to-create-a-secure-zip-

Is there an example of the actual code / function calls required to actually create one from a list of files as I've searched and manually looked in various files and can't find anything?
Tx

**Reply by nmoinvaz** [reply]
I have removed it in 4.0.

**Comment by dgm3333** [comment]
Ahh.
I'm working my way though the console application code which is actually pretty clear once you back track it.
(and figure out that you need to include mz_strm.h before mz_zip_rw.h as mz_zip_rw requires mz_stream_write_cb to be defined but doesn't correctly include mz_strm.h)

This has defines, although unfortunately neither defines nor enums are used in the primary code, so you have to keep your fingers crossed you've guessed the correct one for a particular parameter
https://github.com/zlib-ng/minizip-ng/blob/master/mz.h

This has the option definition
https://github.com/zlib-ng/minizip-ng/blob/master/minizip.c#L25

This has code for setting passwords when a zip job is added
https://github.com/zlib-ng/minizip-ng/blob/master/minizip.c#L254

This is where main() takes the aes/pwd/etc parameters
https://github.com/zlib-ng/minizip-ng/blob/master/minizip.c#L605
